### PR TITLE
Adds Query::executeAndClose() method

### LINKED ISSUE #10008: Add rowCount() to Query to simplify code.
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.11
* Platform and Target: Apache2, Ubuntu, MySQL, PHP 7.1

### What you did

Working with queries like DELETE/UPDATE creates extra work for the developer to simply verify the modified row count.

```
   $statement = $this->find()
           ->update()
           ->where([.....])
           ->execute();

   try {
         return $statement->rowCount();
   } finally {
         $statement->closeCursor();
   }
```

It's recommended to close the cursor when you're done with the statement. 

### What happened

You can do the same above, but skip closing the cursor.

```
   return $this->find()
           ->update()
           ->where([.....])
           ->execute()
           ->rowCount();
```

This leaves the cursor to be handled by gc..

### What you expected to happen

I think we could add a `rowCount()` method to the Query object to simplify this for people who want to always close the cursor.

```
   return $this->find()
           ->update()
           ->where([.....])
           ->rowCount();
```

The above would execute the query, return the count and close the statement.

### PR DESCRIPTION
**Pull request by thinkingmedia**
Closes #10008 

Adds a method that returns row count and closes the statement.

### CODE REVIEW

**Comment by saeideng** [comment]
ExecuteAndCloseWithRowCount() :smile: 
Can we use rowCount() for every query?

**Comment by thinkingmedia** [comment]
@saeideng see the comments in the code. rowCount will return 0 for SELECT statements.

File: src/Database/Query.php [hunk 1]
```diff
@@ -226,6 +226,37 @@ public function execute()
...
+     *                 ->where(['published'=>false])
+     *                 ->executeAndClose();
+     * ```
+     *
+     * The above example will change the published column to true for all false records, and return the number of
+     * records that were updated.
+     *
+     * @return int
+     */
+    public function executeAndClose()
```
Comment (markstory) [inline src/Database/Query.php]: rowCountAndClose() :bike: 🏠

Comment (thinkingmedia) [inline src/Database/Query.php]: I had `rowCount()` disliked, see comments.
https://github.com/cakephp/cakephp/issues/10008

**Comment by saeideng** [comment]
This make confuse between update/delete and select 
So i am :-1:

File: src/Database/Query.php (hunk 1 above)
Comment (saeideng) [inline src/Database/Query.php]: ApplyAndClose()

**Comment by thinkingmedia** [comment]
@saeideng if enough people don't like it, then I'll close the PR, but the issue was open for a long time.

This is just a convenience method for people that do a lot of custom update queries. Otherwise you have extra lines of code just to call rowCount() and close the statement.

**Comment by saeideng** [comment]
I unhappy with  this name 
But if you find a good name I'll :+1:

**Comment by saeideng** [comment]
Now i thing/looking this method name(execute) is for update/delete / insert and select

**Comment by thinkingmedia** [comment]
# New Method Name

Please post your recommendations for a new method name, and vote for the method name you like.

File: src/Database/Query.php (hunk 1 above)
Comment (inoas) [inline src/Database/Query.php]: `public function finishHim()` @ https://www.youtube.com/watch?v=SV6JYR9PwIY

Well this is what it does? `executeAndCloseStatementThenMaybeReturnCount()`

`executeReturnCountClose()` maybe but I do like `executeAndClose()` tbh

**Comment by inoas** [comment]
`ExecuteAndCloseWithRowCount()` may be verbose but LGTM

**Comment by saeideng** [comment]
RowCountAndClose()
ChangeAndClose()
ApplyAndClose()

**Comment by markstory** [comment]
RowCountAndClose sounds good to me. You can't get a row count without executing a query.

**Comment by saeideng** [comment]
also you need an error/exception  for select queries

**Comment by thinkingmedia** [comment]
@saeideng I don't want to throw an exception on a select query. That is logic that should be handled by the database driver. You can already call `execute()->rowCount()` without an error. If there is to be an error it should be in the `rowCount()` method at the driver level.

Maybe there is a database driver that supports row counts on selects.

**Comment by saeideng** [comment]
Ok

**Comment by thinkingmedia** [comment]
okay, this has been updated from comments.

**Pull request merged**